src/mbgl/style/expression/distance.cpp

```diff
@@ -0,0 +1,601 @@
+            [](const auto&) -> double { return std::numeric_limits<double>::quiet_NaN(); });
+}
+
+struct Arguments {
+    Arguments(GeoJSON& geojson_, mapbox::cheap_ruler::CheapRuler::Unit unit_)
+        : geojson(std::move(geojson_)), unit(unit_) {}
+
+    GeoJSON geojson;
+    mapbox::cheap_ruler::CheapRuler::Unit unit;
+};
+
+optional<Arguments> parseValue(const style::conversion::Convertible& value, style::expression::ParsingContext& ctx) {
+    if (isArray(value)) {
+        // object value, quoted with ["Distance", GeoJSONObj, "unit(optional)"]
+        auto length = arrayLength(value);
+        if (length != 2 && length != 3) {
+            ctx.error("'distance' expression requires either one argument or two arguments, but found " +
+                      util::toString(arrayLength(value) - 1) + " instead.");
+            return nullopt;
+        }
+
+        // Parse Unit info for distance calculation, "Meters" by default
+        mapbox::cheap_ruler::CheapRuler::Unit unit = mapbox::cheap_ruler::CheapRuler::Unit::Meters;
+        if (length == 3) {
+            auto input = toString(arrayMember(value, 2));
+            if (input == nullopt) {
+                ctx.error("Failed to parse unit argument from 'distance' expression");
+                return nullopt;
+            }
+            if (*input == "Meters" || *input == "Metres") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Meters;
+            } else if (*input == "Kilometers") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Kilometers;
+            } else if (*input == "Miles") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Miles;
+            } else if (*input == "NauticalMiles") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::NauticalMiles;
+            } else if (*input == "Yards") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Yards;
+            } else if (*input == "Feet") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Feet;
+            } else if (*input == "Inches") {
+                unit = mapbox::cheap_ruler::CheapRuler::Unit::Inches;
+            } else {
+                ctx.error(
+                    "'distance' expression only accepts following Units:  'Kilometers', 'Miles', 'NauticalMiles', "
+                    "'Meters', 'Metres', 'Yards', 'Feet', 'Inches'.");
+                return nullopt;
+            }
+        }
+        // Parse geometry info
+        const auto& argument1 = arrayMember(value, 1);
+        if (isObject(argument1)) {
+            style::conversion::Error error;
+            auto geojson = toGeoJSON(argument1, error);
+            if (geojson && error.message.empty()) {
+                return Arguments(*geojson, unit);
+            }
+            ctx.error(error.message);
+        }
+    }
+    ctx.error(
+        "'distance' expression needs to be an array with format [\"Distance\", GeoJSONObj, \"unit(optional, Meters by "
```

Comment:
`unit(optional)` is a bit misleading, looks like special function, like `rgba(r,g,b,a)`, however It is just a string.